scapy/contrib/isotp.py

```diff
@@ -1611,3 +1613,368 @@ def recv(self, x=0xffff):
+                               communication on the bus
+            sniff_time: time the scan waits for isotp flow control responses
+                        after sending a first frame
+            output_format: defines the format of the returned
+                           results (text, code or sockets). Provide a string
+                           e.g. "text". Default is "socket".
+            can_interface: interface used to create the returned code/sockets
+            verbose: displays information during scan
+
+    Scan for ISOTP Sockets in the defined range and returns found sockets
+    in a specified format. The format can be:
+    text: human readable output
+    code: python code for copy&paste
+    sockets: if output format is not specified, ISOTPSockets will be
+             created and returned in a list
+    """
+
+    if verbose:
+        print("Filtering background noise...")
+
+    # Send dummy packet. In most cases, this triggers activity on the bus.
+    dummy_pkt = CAN(identifier=0x123,
+                    data=b'\xaa\xbb\xcc\xdd\xee\xff\xaa\xbb')
+
+    background_pkts = sock.sniff(timeout=noise_listen_time,
+                                 started_callback=lambda:
+                                 sock.send(dummy_pkt))
+
+    noise_ids = list(set([pkt.identifier for pkt in background_pkts]))
+
+    if extended_addressing:
+        found_packets = scan_extended(sock, scan_range, noise_ids=noise_ids,
+                                      sniff_time=sniff_time, verbose=verbose)
+    else:
+        found_packets = scan(sock, scan_range, noise_ids=noise_ids,
+                             sniff_time=sniff_time, verbose=verbose)
+
+    filter_periodic_packets(found_packets, verbose)
+
+    if output_format == "text":
+        return generate_text_output(found_packets)
+    elif output_format == "code":
+        return generate_code_output(found_packets, can_interface)
+    else:
+        return generate_isotp_list(found_packets, can_interface)
+
+
+def generate_text_output(found_packets):
+    if len(found_packets) == 0:
+        return "No packets found."
+    else:
+        text = "\nFound " + str(len(found_packets)) + \
+               " ISOTP-FlowControl Packet(s):"
+        for pack in found_packets:
+            extended_id = pack > 0x7ff
+            if extended_id:
+                send_id = int(pack / 256)
+                send_ext = pack - (send_id * 256)
+                ext_id = hex(orb(found_packets[pack][0].data[0]))
+                text += "\nSend to ID:             " + hex(send_id) + \
+                        "\nSend to extended ID:    " + hex(send_ext) + \
+                        "\nReceived ID:            " + \
+                        hex(found_packets[pack][0].identifier) + \
+                        "\nReceived extended ID:   " + ext_id + \
+                        "\nMessage:                " + \
+                        found_packets[pack][0].__repr__()
```

Comment:
It's cleaner to `repr(x)` than `x.__repr__()`

Reply:
done